Add User.changePassword(id, old, new, cb)

bajtos opened the pull request:
### Description

Implement a new method for changing user passwords the secure way.
The method requires the old password to be provided before a new
password can be used.

JavaScript API:

    User.changePassword(userId, oldPassword, newPassword[, cb])

REST API:

    POST /api/users/change-password
    Authorization: your-token-id
    Content-Type: application/json

    {"oldPassword":"old-pass", "newPassword": "new-pass"}

#### Related issues

<!--
Please use the following link syntaxes:

- #49 (to reference issues in the current repository)
- strongloop/loopback#49 (to reference issues in another repository)
-->

- #382

### Checklist

<!--
Please mark your choice with an "x" (i.e. [x], see
https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments)
-->

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

loay commented on common/models/user.js:
```diff
@@ -354,6 +354,60 @@ module.exports = function(User) {
+        const err = new Error(`User ${userId} not found`);
+        Object.assign(err, {
+          code: 'USER_NOT_FOUND',
+          statusCode: 400,
```
statusCode should be 404 to be conform with the rest of the code example https://github.com/strongloop/loopback/blob/9c904629d180e3977622d04c80ea822d69282f6b/common/models/user.js#L621

superkhau replied:
Actually, I'm not so sure about 404 now. 401 seems more correct according to SO: http://stackoverflow.com/questions/1959947/whats-an-appropriate-http-status-code-to-return-by-a-rest-api-service-for-a-val

bajtos replied:
Good catch. I was thinking about 404 too, but it seemed inappropriate to me.

> The server has not found anything matching the Request-URI.

The Request URI is `/api/users/change-password`, which is a valid existing path.

The problem is in the access token supplied in the Authorization header, because the user authenticated by this token no longer exists.

On the second thought, I agree with @superkhau that 401 is a more appropriate response status code.

loay commented on test/user.test.js:
```diff
@@ -1267,6 +1267,109 @@ describe('User', function() {
+          const props = Object.assign({}, err);
+          expect(props).to.contain({
+            code: 'USER_NOT_FOUND',
+            statusCode: 400,
```
This should be 404... as per the description of the unit test: 'fails with 404 for unknown user id'

superkhau commented on test/user.test.js:
```diff
@@ -1267,6 +1267,109 @@ describe('User', function() {
+        });
+    });
+
+    it('forwards "options" argument', () => {
```
forwards the

ebarault commented:
I understand the original issue as: "prevent someone that has stolen a valid access token to modify the related user's password".

So actually i'm not sure if there's a good reason not to make this new method an instance method, with an ACL restricted to `$owner`. This would enforce the double check: the user is logged in and knows the current password.

In the case the user is not logged in because he lost his password, then resetPassword should be used, with the caveat addressed that currently tokens emitted by resetPassword are valid application wide.

bajtos commented on common/models/user.js:
```diff
@@ -812,6 +866,22 @@ module.exports = function(User) {
+          {arg: 'newPassword', type: 'string', required: true, http: {source: 'form'}},
+          {arg: 'options', type: 'object', http: 'optionsFromRequest'},
+        ],
+        http: {verb: 'POST', path: '/change-password'},
```
We typically use `POST` for non-RESTful actions, see e.g. `login`, `logout` or `resetPassword`.

bajtos commented:
@ebarault thank you for chiming in.

> prevent someone that has stolen a valid access token to modify the related user's password

Exactly, that's the problem we need to address IMO.

> + the extra comment: "This method should have ACLs configured to allow only the owner to execute it"

I think that's an implementation detail that can be changed as long as the root cause is fixed.

> So actually i'm not sure if there's a good reason not to make this new method an instance method, with an ACL restricted to $owner. This would enforce the double check: the user is logged in and knows the current password.

IMO, the current implementation is equally secure as using `$owner` ACL, because we are modifying the currently logged in user as provided by the access token used to authenticate the request.

My objections against making `changePassword` a prototype method stem from the fact that as a prototype method, it would be exposed on `/api/users/:id/change-password`. This means:
 1. Clients would have to build this URL dynamically. My proposal allows them to always use the same static URL.
 2. The request contains two fields that may be in conflict (accessToken may authenticate a different user from the one specified via the instance `id` argument). I prefer to design the API in  a way that prevents this situation, instead of having to handle it in our code.
 3. It's no secret that the performance of our ACL system is not great, the $owner check comes with performance cost that I would rather avoid.

bajtos commented:
I have changed the controversial status code to 401, see https://github.com/strongloop/loopback/pull/3299#discussion_r107646218 and addressed most of the comments.

The remaining one is PUT vs POST discussion - see https://github.com/strongloop/loopback/pull/3299#discussion_r107551212

ebarault commented:
> IMO, the current implementation is equally secure as using $owner ACL, because we are modifying the currently logged in user as provided by the access token used to authenticate the request.
👍 

May i suggest we add a prototype method shortcut not exposed over REST for further use for example in building different endpoints reusing this code?

bajtos commented:
> May i suggest we add a prototype method shortcut not exposed over REST for further use for example in building different endpoints reusing this code?

@ebarault good idea! Added in 2c0ab9e, PTAL

loay commented:
I created a PR to fix failing CI for `gateway-director-bluemix@develop`. I will let you guys know when it is landed.

bajtos commented on test/user.test.js:
```diff
@@ -1267,6 +1267,109 @@ describe('User', function() {
+        err => {
+          // workaround for chai problem
+          //   object tested must be an array, an object, or a string,
+          //    but error given
```
FWIW, I have fixed the indentation in 4e60e62, I don't know why github did not recognize this comment as outdated.

safwen1 commented:
Why does this method delete current accessToken ?

bajtos commented:
> Why does this method delete current accessToken ?

@safwen1 please open a new github issue to discuss this problem and mention my handle (`@bajtos`) so that I get an email notification.
